
<!-- README.md is generated from README.Rmd. Please edit that file -->

# isoswitch

<!-- badges: start -->

<!-- badges: end -->

The goal of isoswitch is to facilitate the characterization of isoform
expression in long-read single-cell datasets. It includes a set of
functions and reports built on top of `Seurat`, `ggplot` and `rmarkdown`
that can be used to search, visualize and document isoform expression
patterns, and particularly isoform switches between cell identities.

## Installation

You can install the development version of isoswitch from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("atienza-ipmc/isoswitch")
```

## General Workflow

1.  Input data setup & pre-processing
2.  Isoform characterization
3.  Isoform switch detection
4.  Gene reports

Below is a short overview of the package functionality:

### 1 - Input data setup & pre-processing

Isoswitch is designed to work with `Seurat` objects with gene- and
isoform-level counts.

In particular, the ScNaUmi-seq protocol (Lebrigand et al 2020) generates
two  
count matrices that need to be loaded into respective Seurat assays
before starting the analysis, more specifically:

  - A gene-level \[gene x cell\] matrix count, generated by 10X
    CellRanger pipeline. Typically stored in assay “RNA” in Seurat
    object.
  - An isoform-level \[isoform x cell\] matrix count generated by
    SiCeLoRe pipeline, stored in a separate “isoform” assay.

> **Note**  
> isoswith expects the row names of the isoform count matrix to follow
> the format “\[gene\_name\]..\[transcript\_id\]”, example:
> “BCS1L..ENST00000359273”

``` r
head(rownames(seurat@assays$multi@counts))
#> [1] "BCS1L..ENST00000359273"   "PPP1R10..ENST00000461593"
#> [3] "OSBPL9..ENST00000428468"  "TEF..ENST00000406644"    
#> [5] "CBX5..ENST00000209875"    "TRAP1..ENST00000246957"
```

After loading up gene and isoform counts on the Seurat object, the
method `iso_preprocess()` prepares the isoform matrix by removing
low-expression transcripts as well as removing single-isoform genes
which are irrelevant for the isoform switch analysis.

The resulting “multi-isoform” matrix is stored as a new assay on the
input Seurat object.

``` r
seurat <- iso_preprocess(seurat, assay="ISO", new_assay="multi", filter_threshold=5)
```

### 2\. Isoform characterization

The method `iso_compute_stats()` parses the isoform raw count matrix
returning a data frame with stats on the expression of each transcript

  - `feature`, `gene_id`, `transcript_id` gene/transcript identifiers
  - `sum` Total UMI counts for the transcript
  - `total_gene` Total UMI counts for the gene
  - `n_isofs` Number of distinct isoforms detected for the gene
  - `max_sum` Max sum value for the isoform with the highest expression
  - `perc` Relative percentage of isoform UMI count vs gene total.
  - `is_major` (Boolean) is this considered a major isoform
  - `is_top` (Boolean) is this highest expressed isoform

<!-- end list -->

``` r
stats <- iso_compute_stats(seurat@assays$multi@counts) %>% arrange(gene_id)
head(stats, n=4)
#>                 feature gene_id   transcript_id sum total_gene n_isofs max_sum
#> 1 A1BG..ENST00000596924    A1BG ENST00000596924   5          8       2       5
#> 2 A1BG..ENST00000598345    A1BG ENST00000598345   3          8       2       5
#> 3  A2M..ENST00000495709     A2M ENST00000495709  10         14       2      10
#> 4  A2M..ENST00000318602     A2M ENST00000318602   4         14       2      10
#>       perc is_major is_top
#> 1 62.50000     TRUE   TRUE
#> 2 37.50000     TRUE  FALSE
#> 3 71.42857     TRUE   TRUE
#> 4 28.57143    FALSE  FALSE
```

The method `plot_assay_stats()` builds on this data to plot a summary
with number of genes, number of transcripts, distribution of isoforms
and number of genes per cell type that can be used to describe succintly
the isoform distribution in the dataset.

``` r
plot_assay_stats(seurat, "isoform")
```

![alt text](./man/figures/Fig4_isosummary.png)

### 3\. Isoform switch search

The term “isoform switch” refers to an event where two isoforms of the
same gene are considered markers of different clusters.

The marker search is implemented on the method `ISO_SWITCH_ALL()`. Any
extra parameters are passed on to the underlying Seurat’s FindMarkers
call to fine tune the search space.

``` r
clusters <- levels(seurat@active.ident)
switch_markers <- ISO_SWITCH_ALL(seurat, clusters, assay="isoform", 
                                 min.pct=0, logfc.threshold=0.40)
```

`ISO_SWITCH_ALL()` returns a data frame of transcripts identified as
markers of a given cluster for a given gene, one transcript per row.

The method `compute_switches()` takes a list of markers as input and
combines them into a data frame where all the parameters defining an
isoform switch is contained in a single row.

``` r
switches <- compute_switches(switch_markers)
```

The helper method `gene_switch_table()` can be used to print a summary
of isoform switches for reports.

``` r
gene_switch_table(switch_markers)
```

<div id="htmlwidget-f3490ee769c955969bba" class="reactable html-widget" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-f3490ee769c955969bba">{"x":{"tag":{"name":"Reactable","attribs":{"data":{"geneId":["HYAL2","RPS24","MYL6","EEF1B2","LYPLA2","LGALS9","RTN4","VAMP5","GCHFR","LMNA","EIF3C","HSPA8","COMT","TPM4","PIGT","SAT1","SLC25A3","EDF1","CAPZB","ATP5F1B","CLU","GNAS","UBAC1","ATOX1","CENPM","PDLIM1","UBA52","SMIM5","GSN","CTSH","HLA-C","SPINT1-AS1","NFE2L2","VGLL4","VCP","EIF6","MRPS21","ROMO1","MUC20","GMDS","DUT","BCAP31","LIMS1","ACTG1","SAP18","SLC22A18","GPX4","ADAM15","POLR1D","RPS23","UBE2D3","DMKN","IFI27","MGMT","RPL37A","ELOC","AURKAIP1","PPCS","DDOST","ANAPC11","KRT10","RPL9","ARPC4","TPI1","HNRNPA3","RPS6","RPS9","GSTM2","FXYD3","ELF3","MDH1","CUTA","GUK1","IRF9","SPSB3","SSNA1","ANKRD37","ANXA11","PSMD8","RPL10A","SERBP1","RAC1","TIMM8B","RBP1","ENSA","NAA20","DYNLL1","SYTL1","SINHCAF","NDUFB10","TMA7","CENPS","SNU13","ENY2","HERPUD1","RBIS","CAT","SSBP2","C18orf32","PSMD13","BOLA3","IFT57","KRAS","SMAGP","EIF3D","DRG1","CTNNA1","OAT","GARS","RBBP4","RPS18","PKM","LAMTOR5","BTF3","MED27","MRPL22","CSDE1","TAP1","PFDN5","ST6GALNAC1","ALDH3A1","REXO2","NDUFB11","DNAJC17","NDUFB1","ANXA3","SAMM50","ZNF394","DEF8","CCT2","SF3B1","PSMB8","PDLIM5","LSM7","MTG1","PDIA4","ELMO1","ENKD1","MFSD11","NRAV","PER1"],"avg_log2FC":[2.7271453858888,-2.86561572097228,1.57196436962651,-1.10690154640962,1.01731100398389,1.13599546110602,-1.15168669082671,2.82810597398475,0.583844734187208,1.05295836245785,0.815517481657346,-0.674317552384904,0.69591695542917,1.71072363042502,0.759887780501296,0.765259077791314,-0.43798916872275,-0.928139791217959,0.751128934096753,-0.737007869414109,1.34605407345589,1.14593581935895,4.21132056946097,1.00697656997552,-1.32142102905414,1.22290689279748,-2.12959138254209,-1.1736602067795,1.11578023184261,-1.18575313555338,1.62549899322714,0.477811699614975,1.55530237149189,1.49402931356928,1.09197687092979,0.674854780339262,-1.44565095196998,2.2734992918632,0.76389129985952,1.26157173487636,-1.888633172001,1.04394721896959,0.888346563114107,-1.31840928048613,0.636085657839371,0.529544939382764,-1.23980861850817,0.712937882487463,-0.451130677025764,0.4545510743121,0.409870539154288,0.894267793051611,-2.06241248201347,-4.02896324183046,1.67717304340649,0.4209921559334,-0.498766866993651,1.52513725672289,0.650249227577255,-1.46310009128072,-2.51043099565993,0.434702079099451,3.00726308157291,-1.23217015327233,-0.666124321570537,0.651775020019353,0.604754121688295,3.64272879585594,0.834522370505686,0.837957622941428,-1.45277510309773,-0.524722501871182,-1.97957002901982,-4.31239231490676,2.60015380993155,-0.832974560880359,-3.32291863010711,1.76262947555145,1.5842040521305,1.30490890559107,0.979878885685496,-0.534985810401863,-0.424838165343966,-0.447571462396223,0.64052810748988,0.566025915106785,-1.53804483021815,1.28486475625088,0.779737818326529,-0.511219717820816,1.62856179185454,-0.597825238885421,-1.20198154425021,-0.445296760223896,0.628627451452616,-0.550567708696525,0.508267052793914,3.39362539665878,-0.75760592899798,0.522033334479054,0.440436571212016,3.11003193064754,-1.10156552302089,1.29426072975104,0.582380420713453,-0.537882574447879,0.620873681454558,0.595597675751656,0.553989156946767,2.55971765606657,0.4885238377232,-1.13422681946477,-1.63308925876458,0.668354036908783,-2.14902249027701,-0.914542347588017,0.482833072221763,0.47822286031122,-0.819985784136938,-0.569511026028128,-0.639753718425936,0.474133138659535,-0.535236782874103,0.802995340046304,-0.639706304584193,-0.892792213990508,0.520823071776639,2.72138042399,-0.604475444242252,0.728078881675774,3.02086407888324,-0.910139161294689,3.27440593478457,-0.59728691577856,-0.490146865249793,0.450428949802628,-0.596839832643237,1.84966572691557,-1.15024263558061,-0.49771995470749,0.784194291085175],"pct.1":[0.345,0.53,0.977,0.893,0.071,0.106,0.16,0.619,0.305,0.288,0.254,0.554,0.7,0.274,0.169,0.655,0.772,0.156,0.254,0.277,0.322,0.373,0.25,0.837,0,0.441,0.104,0.006,0.458,0.071,0.678,0.324,0.339,0.124,0.288,0.757,0.06,0.637,0.232,0.219,0.029,0.271,0.237,0.695,0.626,0.102,0.04,0.135,0.045,0.824,0.149,0.173,0.02,0.006,0.558,0.347,0.792,0.436,0.182,0.021,0,0.294,0.083,0.104,0.04,0.681,0.709,0.188,0.237,0.458,0,0.062,0,0.014,0.104,0.356,0.002,0.479,0.474,0.565,0.203,0.027,0.042,0.04,0.11,0.157,0,0.089,0.169,0.04,0.409,0.009,0.059,0.067,0.102,0.088,0.102,0.125,0.02,0.119,0.08,0.083,0.015,0.168,0.136,0.045,0.237,0.153,0.119,0.104,0.22,0.044,0.021,0.1,0,0.097,0.076,0.085,0.292,0.004,0.013,0.391,0.04,0.11,0.015,0.034,0.119,0.083,0,0.186,0.125,0.015,0.125,0.044,0.013,0.052,0,0.021,0,0,0.029],"pct.2":[0.001,0.927,0.777,0.96,0.002,0.004,0.432,0.193,0.116,0.029,0.025,0.662,0.359,0.054,0.012,0.406,0.837,0.615,0.029,0.685,0.05,0.066,0.028,0.118,0.281,0.07,0.827,0.103,0.105,0.345,0.238,0.045,0.056,0.006,0.04,0.08,0.881,0.021,0.018,0.013,0.532,0.054,0.03,0.89,0.04,0.002,0.625,0.029,0.271,0.146,0.032,0,0.562,0.125,0.02,0.094,0.14,0.015,0.009,0.661,0.403,0.067,0.006,0.629,0.322,0.125,0.125,0.004,0.04,0.155,0.375,0.373,0.45,0.146,0.011,0.476,0.062,0.021,0.021,0.083,0.034,0.182,0.515,0.5,0.029,0.013,0.336,0.011,0.023,0.208,0,0.145,0.525,0.255,0.007,0.61,0.014,0.018,0.508,0.018,0.556,0.008,0.343,0.055,0.016,0.189,0.052,0.022,0.013,0.014,0.843,0.458,0.416,0.571,0.042,0.362,0.014,0.005,0.814,0.117,0.169,0.02,0.431,0.004,0.356,0.169,0.014,0.005,0.102,0.039,0.004,0.296,0.004,0.351,0.125,0.005,0.015,0,0.02,0.119,0],"p_val_adj":[3.86e-229,4.34e-173,7.75e-147,1.3e-75,3.89e-34,1.84e-32,9.96e-29,4.94e-28,1.06e-25,5.61e-23,1.75e-20,1.81e-20,2.84e-19,5.16e-19,2.82e-18,3.19e-18,6.93e-18,1.07e-17,5.36e-17,4.54e-15,5.4e-15,9.23e-15,1.05e-14,1.06e-14,4.18e-14,1.57e-13,3.23e-13,1.72e-12,3.7e-12,1.53e-11,9.04e-11,1.53e-10,1.05e-09,1.58e-09,3.33e-09,4.78e-09,4.91e-09,5.68e-09,1.17e-08,2.45e-08,3.19e-08,3.98e-08,6.92e-08,1.42e-07,1.68e-07,1.83e-07,2.31e-07,2.87e-07,4.78e-07,4.83e-07,7.5e-07,8.44e-07,1.04e-06,1.12e-06,1.14e-06,2.22e-06,3.09e-06,3.73e-06,4.03e-06,4.07e-06,5.89e-06,7.46e-06,1.22e-05,1.29e-05,1.35e-05,1.56e-05,1.98e-05,2.51e-05,3.13e-05,3.38e-05,3.95e-05,7.07e-05,9.08e-05,0.000122,0.000194,0.000198,0.000218,0.00023,0.000318,0.000351,0.00042,0.000471,0.000494,0.000511,0.000532,0.000534,0.00057,0.000659,0.000696,0.000762,0.00083,0.000866,0.00135,0.0014,0.00143,0.00148,0.00148,0.00162,0.0018,0.00182,0.00201,0.00218,0.00219,0.00226,0.00333,0.00348,0.00422,0.00518,0.00564,0.00581,0.0062,0.00624,0.00725,0.00751,0.00818,0.00853,0.00861,0.00861,0.0087,0.0105,0.013,0.0169,0.0185,0.0188,0.0198,0.0255,0.0263,0.0298,0.0374,0.0395,0.0406,0.0409,0.0421,0.0422,0.0457,0.0474,1,1,1,1,1],"contrast":["EC Venous vs Suprabasal_1","Squamous vs Goblet","Goblet vs Basal","Squamous vs Basal","Multiciliated vs Suprabasal_1","EC Venous vs Suprabasal_1","Basal vs Suprabasal_1","EC Venous vs Squamous","Squamous vs Suprabasal_1","Deuterosomal vs Suprabasal_1","Deuterosomal vs Suprabasal_1","Goblet vs Suprabasal_1","Squamous vs Basal","EC Venous vs Suprabasal_1","Deuterosomal vs Suprabasal_1","Squamous vs Suprabasal_1","Goblet vs Suprabasal_1","Multiciliated vs Cycling basal","Deuterosomal vs Suprabasal_1","Multiciliated vs Basal","Deuterosomal vs Suprabasal_1","Deuterosomal vs Suprabasal_1","Serous vs Suprabasal_1","Squamous vs T","Multiciliated vs Cycling basal","Deuterosomal vs Basal","Serous vs Cycling basal","Goblet vs Multiciliated","Deuterosomal vs Suprabasal_1","Multiciliated vs Suprabasal_1","Deuterosomal vs Basal","Squamous vs Multiciliated","Deuterosomal vs Basal","EC Venous vs Squamous","Deuterosomal vs Basal","Squamous vs SMG Goblet","SMG Goblet vs Deuterosomal","Squamous vs Serous","Goblet vs Multiciliated","Goblet vs Multiciliated","EC Arterial vs Cycling basal","Deuterosomal vs Suprabasal_1","Deuterosomal vs Cycling basal","Deuterosomal vs Basal","Squamous vs SMG Goblet","Deuterosomal vs Basal","SMG Goblet vs Suprabasal_2","Goblet vs Basal","Multiciliated vs Cycling basal","Squamous vs Serous","Squamous vs Basal","Squamous vs Multiciliated","SMG Goblet vs Suprabasal_2","Squamous vs Serous","Squamous vs SMG Goblet","Squamous vs Multiciliated","Squamous vs SMG Goblet","Squamous vs T","Squamous vs Multiciliated","Serous vs Deuterosomal","EC Arterial vs Squamous","Squamous vs Multiciliated","Serous vs Suprabasal_1","Serous vs Basal","Multiciliated vs Deuterosomal","Squamous vs Serous","Squamous vs Serous","Serous vs Multiciliated","Deuterosomal vs Basal","Deuterosomal vs Suprabasal_1","EC Arterial vs Squamous","Multiciliated vs Deuterosomal","SMG Goblet vs Cycling basal","Squamous vs Serous","Serous vs Suprabasal_1","Squamous vs Cycling basal","Goblet vs Serous","Squamous vs Serous","Squamous vs Serous","Squamous vs Serous","Deuterosomal vs Basal","Multiciliated vs Basal","Serous vs Cycling basal","SMG Goblet vs Cycling basal","Goblet vs Basal","Squamous vs Multiciliated","EC Arterial vs Cycling basal","Multiciliated vs Basal","Deuterosomal vs Basal","Multiciliated vs Cycling basal","Squamous vs SMG Goblet","Multiciliated vs Cycling basal","T vs Deuterosomal","Multiciliated vs Cycling basal","Deuterosomal vs Basal","EC Arterial vs Deuterosomal","Deuterosomal vs Suprabasal_1","Serous vs Suprabasal_1","SMG Goblet vs Deuterosomal","Deuterosomal vs Suprabasal_1","SMG Goblet vs Cycling basal","Serous vs Suprabasal_1","EC Arterial vs Squamous","EC Venous vs Suprabasal_1","Deuterosomal vs Basal","Multiciliated vs Suprabasal_1","Deuterosomal vs Suprabasal_2","Deuterosomal vs Basal","Deuterosomal vs Basal","Serous vs Suprabasal_1","SMG Goblet vs Suprabasal_2","EC Arterial vs Deuterosomal","Serous vs Cycling basal","SMG Goblet vs Cycling basal","Squamous vs Serous","EC Venous vs Goblet","Goblet vs Basal","Deuterosomal vs Basal","EC Venous vs Deuterosomal","Multiciliated vs Suprabasal_2","Multiciliated vs Deuterosomal","Squamous vs SMG Goblet","SMG Goblet vs Basal","Goblet vs Multiciliated","EC Arterial vs Deuterosomal","Squamous vs Deuterosomal","Deuterosomal vs Basal","Serous vs Basal","Multiciliated vs Deuterosomal","Deuterosomal vs Cycling basal","Serous vs Multiciliated","T vs Cycling basal","Serous vs Multiciliated","EC Arterial vs Squamous","Multiciliated vs Cycling basal","Goblet vs Basal","EC Arterial vs T","Serous vs T","EC Arterial vs SMG Goblet","Serous vs Deuterosomal","EC Arterial vs SMG Goblet"]},"columns":[{"accessor":"geneId","name":"geneId","type":"character","minWidth":80,"filterable":true,"width":90},{"accessor":"avg_log2FC","name":"avg_log2FC","type":"numeric","minWidth":80,"format":{"cell":{"digits":2},"aggregated":{"digits":2}},"width":90},{"accessor":"pct.1","name":"pct.1","type":"numeric","minWidth":80,"width":70},{"accessor":"pct.2","name":"pct.2","type":"numeric","minWidth":80,"width":70},{"accessor":"p_val_adj","name":"p_val_adj","type":"numeric","minWidth":80,"width":120,"align":"right"},{"accessor":"contrast","name":"contrast","type":"character","minWidth":150,"filterable":true}],"defaultPageSize":15,"paginationType":"numbers","showPageInfo":true,"minRows":1,"striped":true,"compact":true,"style":{"fontFamily":"Work Sans, sans-serif","fontSize":"12px"},"theme":{"color":"#333333","cellPadding":6,"tableStyle":{"fontSize":12},"headerStyle":{"color":"#333333","fontSize":13},"groupHeaderStyle":{"color":"#333333","fontSize":13},"cellStyle":{"display":"flex","flexDirection":"column","justifyContent":"center"}},"dataKey":"4efcf7933ccc561a9646110383d192c1","key":"4efcf7933ccc561a9646110383d192c1"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script>

Alternatively, to facilitate the graphical interpretation of the
results:

  - `plot_marker_matrix()` produces a heatmap of number of unique genes
    per contrast between clusters  
  - `plot_marker_score()` produces a volcano plot showing p-values and
    average logFC for each gene with an isoform switch

<!-- end list -->

``` r
pl1 <- plot_marker_matrix(seurat, switch_markers) 
pl2 <- plot_marker_score(adult, switch_markers, facet=FALSE, overlaps=16)
pl1 | pl2 
```

![alt text](./man/figures/Fig7_isoswitch.png) `plot_marker_score()` can
also plot individual volcano plots for each cluster analyzed with the
parameter facet=TRUE

``` r
plot_marker_score(adult, switch_markers, facet=TRUE, ncol=3)
```

![alt text](./man/figures/Fig7_facet.png)

### 4\. Gene reports

After identifying genes of interest, twere are two ways of drilling down
and producing gene-level reports:

  - The `isoswitch_report()` method produces a compact plot of the gene

<!-- end list -->

``` r
isoswitch_report(seurat, "isoform", gene="HYAL2", marker_list=switch_markers) 
```

![alt text](./man/figures/Fig7_hyal2.png) \* The
`render_html_gene_report()` method renders an [html
version](https://www.isomics.eu/reports/fetal/genes/C3_CD36.nb.html) of
the report
