---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)

# test data
library(dplyr)
library(isoswitch)
seurat <- readRDS("/data/10x_data/10x_5psanger/output/adult50.rds")

```

# isoswitch

<!-- badges: start -->
<!-- badges: end -->

The goal of isoswitch is to facilitate the characterization of isoform expression in long-read single-cell datasets
It includes a set of functions and reports built on top of Seurat, ggplot and rmarkdown that can be used to search, visualize and document isoform expression patterns, and particularly isoform switches between cell identities.


## Installation

You can install the development version of isoswitch from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("atienza-ipmc/isoswitch")
```

## General Workflow 

1. Input data / object setup
2. Isoform characterization
3. Switch search
4. Gene reports

Below is a short overview of the package functionality:

### 1 - Input data / object setup

Isoswitch is designed to work with Seurat objects with gene- and isoform-level counts. 
In particular, the ScNaUmi-seq protocol (Lebrigand et al 2020) produces:

- A gene-level [gene x cell] matrix count, generated by 10X CellRanger pipeline. Typically stored in assay "RNA" in Seurat object.
- An isoform-level [isoform x cell] matrix count, generated by SiCeLoRe. Store in a separate "isoform" assay. By convention, the row names of the isoform count matrix follow the format "<gene_name>..<transcript_id> -> "BCS1L..ENST00000359273"

```{r echo=TRUE, warning=FALSE}
head(rownames(seurat@assays$multi@counts))
```


### 2. Isoform characterization

The method **iso_compute_stats** parses the isoform raw count matrix and returns a data frame with the following fields:

```{r echo=TRUE, warning=FALSE}
stats <- iso_compute_stats(seurat@assays$multi@counts) %>% arrange(gene_id)
head(stats, n=4)
```

The method **plot_assay_stats** builds on this data to plot a summary with number of genes, number of transcripts, distribution of isoforms and number of genes per cell type.

```{r, eval=FALSE}
plot_assay_stats(seurat, "isoform")
```

![alt text](./man/figures/Fig4_isosummary.png)

### 3. Isoform switch search

The term “isoform switch” refers to an event where two isoforms of the same gene are considered markers of different clusters. 

The marker search is implemented on **ISO_SWITCH_ALL** which relies on Seurat's FindMarkers functionality. 

```{r, eval=FALSE}
clusters <- levels(seurat@active.ident)
switch_markers <- ISO_SWITCH_ALL(seurat, clusters, assay="isoform", min.pct=0, logfc.threshold=0.40, verbose=TRUE)
```

The result of **ISO_SWITCH_ALL** is a data frame of transcripts considered statistically significant markers of certain clusters. 
The method **compute_switches** takes a list of markers as input and combines them into switches, ranking the list of switches according to different possible criteria.

```{r, eval=FALSE}
switches <- compute_switches(switch_markers)
```

The helper method **format_switch_table** prettifies the switch table for reports:

To facilitate the interpretation of this data, two methods **plot_marker_matrix** and **plot_marker_score** generate plots with 1) heatmap of number of unique genes per contrast between clusters and 2) volcano-like showing p-values and average logFC for each gene with an isoform switch   

```{r, eval=FALSE}
pl1 <- plot_marker_matrix(seurat, switch_markers) 
pl2 <- plot_marker_score(adult, switch_markers, facet=FALSE, overlaps=16)
pl1 | pl2 
```
 ![alt text](./man/figures/Fig7_isoswitch.png)
Alternatively, **plot_marker_score** can also plot faceted individual plots for each cluster analyzed

```{r, eval=FALSE}
plot_marker_score(adult, switch_markers, facet=TRUE, ncol=3)
```
 ![alt text](./man/figures/Fig7_facet.png)

### 4. Gene reports

After identifying genes of interest, **isoswitch_report** produces a detailed report of the isoform switch


```{r, eval=FALSE}
isoswitch_report(seurat, "isoform", gene="HYAL2", marker_list=switch_markers) 
```
 ![alt text](./man/figures/Fig7_hyal2.png)
